commcare_connect/program/helpers.py

```diff
@@ -0,0 +1,56 @@
+from django.db.models import Avg, Count, DurationField, ExpressionWrapper, F, OuterRef, Q, Subquery
+
+from commcare_connect.opportunity.models import UserVisit, VisitValidationStatus
+from commcare_connect.program.models import ManagedOpportunity, Program
+
+
+def get_annotated_managed_opportunity(program: Program):
+    filter_for_valid__visit_date = ~Q(
+        opportunityaccess__uservisit__status__in=[
+            VisitValidationStatus.over_limit,
+            VisitValidationStatus.trial,
+        ]
+    )
+
+    earliest_visits = (
+        UserVisit.objects.filter(
+            opportunity_access=OuterRef("opportunityaccess"),
+        )
+        .exclude(status__in=[VisitValidationStatus.over_limit, VisitValidationStatus.trial])
+        .order_by("visit_date")
+        .values("visit_date")[:1]
+    )
+
+    managed_opportunities = (
+        ManagedOpportunity.objects.filter(program=program)
+        .order_by("start_date")
+        .annotate(
+            workers_invited=Count("opportunityaccess"),
+            workers_passing_assessment=Count(
+                "opportunityaccess__assessment",
+                filter=Q(
+                    opportunityaccess__assessment__passed=True,
+                    opportunityaccess__assessment__opportunity=F("opportunityaccess__opportunity"),
```

Comment:
is this filter necessary? shouldn't the opportunity access guarantee its the right opportunity already?

Reply:
Thanks Removed [4f88e2a](https://github.com/dimagi/commcare-connect/pull/391/commits/4f88e2aa8a84e76736c6954c1e540b7857143581)

Also added some more test.